dg.go

```diff
@@ -265,12 +265,14 @@ func (n *node[NodeType]) Item() NodeType {
 }
 
 func (n *node[NodeType]) ResolutionStatus() ResolutionStatus {
+	// The status can change, so lock to prevent data races.
 	n.dg.lock.Lock()
 	defer n.dg.lock.Unlock()
 	return n.status
```

Comment:
I can confirm that removing the lock does trigger Golang's race detector. I created a test that spawns two goroutines to test this.

```
joconnel@joconnel-mac ~/Documents/projects/arcaflow/dependencies/go-dgraph $ go test ./... --race --run TestDirectedGraph_TwoOrDependenciesMultiThreaded --count 100
ok  	go.arcalot.io/dgraph	1.212s
joconnel@joconnel-mac ~/Documents/projects/arcaflow/dependencies/go-dgraph $ go test ./... --race --run TestDirectedGraph_TwoOrDependenciesMultiThreaded --count 10000
ok  	go.arcalot.io/dgraph	2.386s
joconnel@joconnel-mac ~/Documents/projects/arcaflow/dependencies/go-dgraph $ go test ./... --race --run TestDirectedGraph_TwoOrDependenciesMultiThreaded --count 100000
ok  	go.arcalot.io/dgraph	13.067s
```
```
joconnel@joconnel-mac ~/Documents/projects/arcaflow/dependencies/go-dgraph $ go test ./... --race --run TestDirectedGraph_TwoOrDependenciesMultiThreaded --count 1
==================
WARNING: DATA RACE
Write at 0x00c0000102b0 by goroutine 7:
  go.arcalot.io/dgraph.(*node[go.shape.string]).resolveNode()
      /Users/joconnel/Documents/projects/arcaflow/dependencies/go-dgraph/dg.go:307 +0x258
  go.arcalot.io/dgraph.(*node[go.shape.string]).ResolveNode()
      /Users/joconnel/Documents/projects/arcaflow/dependencies/go-dgraph/dg.go:293 +0xf4
  go.arcalot.io/dgraph.(*node[string]).ResolveNode()
      /Users/joconnel/Documents/projects/arcaflow/dependencies/go-dgraph/dg.go:290 +0x48
  go.arcalot.io/dgraph_test.TestDirectedGraph_TwoOrDependenciesMultiThreaded.func1()
      /Users/joconnel/Documents/projects/arcaflow/dependencies/go-dgraph/dg_test.go:728 +0x9c

Previous read at 0x00c0000102b0 by goroutine 8:
  go.arcalot.io/dgraph.(*node[go.shape.string]).ResolutionStatus()
      /Users/joconnel/Documents/projects/arcaflow/dependencies/go-dgraph/dg.go:271 +0x38
  go.arcalot.io/dgraph.(*node[string]).ResolutionStatus()
      /Users/joconnel/Documents/projects/arcaflow/dependencies/go-dgraph/dg.go:267 +0x20
  go.arcalot.io/dgraph_test.TestDirectedGraph_TwoOrDependenciesMultiThreaded.func2()
      /Users/joconnel/Documents/projects/arcaflow/dependencies/go-dgraph/dg_test.go:733 +0xb4

Goroutine 7 (running) created at:
  go.arcalot.io/dgraph_test.TestDirectedGraph_TwoOrDependenciesMultiThreaded()
      /Users/joconnel/Documents/projects/arcaflow/dependencies/go-dgraph/dg_test.go:726 +0x43c
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.22.2/libexec/src/testing/testing.go:1689 +0x180
  testing.(*T).Run.gowrap1()
      /opt/homebrew/Cellar/go/1.22.2/libexec/src/testing/testing.go:1742 +0x40

Goroutine 8 (running) created at:
  go.arcalot.io/dgraph_test.TestDirectedGraph_TwoOrDependenciesMultiThreaded()
      /Users/joconnel/Documents/projects/arcaflow/dependencies/go-dgraph/dg_test.go:730 +0x55c
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.22.2/libexec/src/testing/testing.go:1689 +0x180
  testing.(*T).Run.gowrap1()
      /opt/homebrew/Cellar/go/1.22.2/libexec/src/testing/testing.go:1742 +0x40
==================
--- FAIL: TestDirectedGraph_TwoOrDependenciesMultiThreaded (0.00s)
    dg_test.go:734: Raced async ready status: false
    dg_test.go:735: Raced async resolution status: waiting
    testing.go:1398: race detected during execution of test
FAIL
FAIL	go.arcalot.io/dgraph	0.201s
FAIL
```
The race detector is working as intended. Obviously, based on the two of your experiences, locking here is not very helpful. But it's a type of race that it's designed to detect.

Reply:
In this case, your test is assuming that asserting `Ready` and `ResolutionStatus` have a "happens-after" relationship with the `Resolve` ... which, the race detector correctly (if unclearly) tells you, is not guaranteed. But that's really beside the point. You're using the lock as a memory barrier to ensure you're reading the values in memory *after* concurrent resolution occurs. That's rarely a useful guarantee in actual code, and, again, if that's really what you want to support than `sync/atomic` would be a vastly more efficient and also a much more clear way to achieve it.

I'm not entirely sure whether the Go race detector is correctly inferring that your code *sequence* (rather than the actual access) is unsafe or whether it's blindly complaining about the unprotected reads (which would be wrong since, on their own, they're fine). I suppose in this case it doesn't matter much, since it's right even if it's right for the wrong reason.